test/integration/master/secrets_enveloping_test.go

```diff
@@ -79,117 +83,203 @@ resources:
-		t.Errorf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
+		e.logger.Errorf("expected %s from KubeAPI, but got %s", secretVal, string(s.Data[secretKey]))
 	}
 }
 
-func startKubeApiWithEncryption(t *testing.T, storageConfig *storagebackend.Config,
-	transformerConfig string) (*kubeapiservertesting.TestServer, error) {
-	tempDir, err := ioutil.TempDir("", "secrets-encryption-test")
-	if err != nil {
-		return nil, fmt.Errorf("failed to create temp directory: %v", err)
+func (e *envelopTest) benchmark(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		_, err := createTestSecret(e.restClient, e.secret.Name+strconv.Itoa(i), e.ns.Name)
+		if err != nil {
+			b.Fatalf("failed to create a secret: %v", err)
+		}
+	}
+}
+
+func (e *envelopTest) getETCDPath() string {
+	return fmt.Sprintf("/%s/secrets/%s/%s", e.storageConfig.Prefix, e.ns.Name, e.secret.Name)
+}
+
+func (e *envelopTest) getKubeAPIServerEncryptionOptions() []string {
+	if e.transformerConfig != "" {
+		return []string{"--experimental-encryption-provider-config", path.Join(e.configDir, encryptionConfigFileName)}
 	}
-	defer os.RemoveAll(tempDir)
 
-	encryptionConfig, err := ioutil.TempFile(tempDir, "encryption-config")
-	if err != nil {
-		return nil, fmt.Errorf("error while creating temp file for encryption config %v", err)
+	return nil
+}
+
+// TestSecretsShouldBeEnveloped is an integration test between KubeAPI and etcd that checks:
+// 1. Secrets are encrypted on write
+// 2. Secrets are decrypted on read
+// when EncryptionConfig is passed to KubeAPI server.
+func TestSecretsShouldBeEnveloped(t *testing.T) {
+	var testCases = []struct {
+		transformerConfigContent string
+		transformerPrefix        string
+		unSealFunc               unSealSecret
+	}{
+		{aesGCMConfigYAML, aesGCMPrefix, unSealWithGCMTransformer},
+		{aesCBCConfigYAML, aesCBCPrefix, unSealWithCBCTransformer},
+		// TODO: add secretbox
 	}
+	for _, tt := range testCases {
+		test, err := newEnvelopeTest(t, tt.transformerConfigContent)
+		if err != nil {
+			test.cleanUp()
+			t.Errorf("failed to setup test for envelop %s, error was %v", tt.transformerPrefix, err)
+			continue
+		}
+		test.run(tt.unSealFunc, tt.transformerPrefix)
+		test.cleanUp()
+	}
+}
+
+// Baseline (no enveloping) - use to contrast with enveloping benchmarks.
+func BenchmarkBase(b *testing.B) {
+	runBenchmark(b, "")
+}
+
+// Identity transformer is a NOOP (crypto-wise) - use to contrast with AESGCM and AESCBC benchmark results.
+func BenchmarkIdentityWrite(b *testing.B) {
+	runBenchmark(b, identityConfigYAML)
+}
+
+func BenchmarkAESGCMEnvelopeWrite(b *testing.B) {
+	runBenchmark(b, aesGCMConfigYAML)
+}
+
+func BenchmarkAESCBCEnvelopeWrite(b *testing.B) {
+	runBenchmark(b, aesCBCConfigYAML)
+}
 
-	if _, err := encryptionConfig.Write([]byte(transformerConfig)); err != nil {
-		return nil, fmt.Errorf("error while writing encryption config: %v", err)
+func runBenchmark(b *testing.B, transformerConfig string) {
+	b.StopTimer()
```

Comment:
why is this necessary?

Reply:
I assume that the question is: why b.StopTimer is necessary.
The idea is to start measurements after test KubeAPI server is online. In other words, the benchmarks should focus on the crypto operations that are triggered by test.benchmark.